src/VisualStudio/IntegrationTests/AbstractEditorTests.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            return new KeyPress(virtualKey, shiftState);
+        }
+
+        protected void DisableSuggestionMode()
+        {
+            _visualStudioWorkspace.UseSuggestionMode = false;
+        }
+
+        protected void EnableSuggestionMode()
+        {
+            _visualStudioWorkspace.UseSuggestionMode = true;
+        }
+
+        protected void InvokeCompletionList()
+        {
+            ExecuteCommand(WellKnownCommandNames.ListMembers);
+            WaitForAsyncOperations(FeatureAttribute.CompletionSet);
+        }
+
+        protected void ExecuteCommand(string commandName)
+        {
+            _visualStudio.Instance.ExecuteCommand(commandName);
+        }
+
+        protected void VerifyCurrentLineText(string expectedText, bool trimWhitespace = true)
+        {
+            var caretStartIndex = expectedText.IndexOf("$$");
+
+            if (caretStartIndex >= 0)
+            {
+                var caretEndIndex = caretStartIndex + "$$".Length;
+
+                var expectedTextBeforeCaret = caretStartIndex < expectedText.Length
+                    ? expectedText.Substring(0, caretStartIndex)
+                    : expectedText;
+
+                var expectedTextAfterCaret = caretEndIndex < expectedText.Length
+                    ? expectedText.Substring(caretEndIndex)
+                    : string.Empty;
+
+                var lineText = _editor.GetCurrentLineText();
+
+                if (trimWhitespace)
+                {
+                    if (caretStartIndex == 0)
+                    {
+                        lineText = lineText.TrimEnd();
+                    }
+                    else if (caretEndIndex == expectedText.Length)
+                    {
+                        lineText = lineText.TrimStart();
+                    }
+                    else
+                    {
+                        lineText = lineText.Trim();
+                    }
+                }
+
+                var lineTextBeforeCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(0, caretStartIndex)
+                    : lineText;
+
+                var lineTextAfterCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(caretStartIndex)
+                    : string.Empty;
+
+                Assert.Equal(expectedTextBeforeCaret, lineTextBeforeCaret);
+                Assert.Equal(expectedTextAfterCaret, lineTextAfterCaret);
+                Assert.Equal(expectedTextBeforeCaret.Length + expectedTextAfterCaret.Length, lineText.Length);
+            }
+            else
+            {
+                var lineText = _editor.GetCurrentLineText();
+
+                if (trimWhitespace)
+                {
+                    lineText = lineText.Trim();
+                }
+
+                Assert.Equal(expectedText, lineText);
+            }
+        }
+
+        protected void VerifyTextContains(string expectedText)
```

Comment:
Indeed, but it's not obvious the method does anything fancier either. Should there just be two methods, one for asserting caret and one that's not? It's not as if the writer of the test doesn't know their intent.